File: shared/typescript/src/web.ts
// web: the web-suite infrastructure shared across every TypeScript server that
// implements it (PLAN §3/§5): the canon constants for /compute and /jwt/sign,
// the shared /validate rules (one zod schema, the single source of truth), the
// SHA-256 compute chain, and the HS256 JWT sign/verify helpers (jose — the house
// pick). Only framework-independent contract values and security policy live
// here; the handlers (header extraction, response shaping, HTML rendering) stay
// per-framework and idiomatic (the §3 idiom boundary).

import { createHash } from "node:crypto";
import { jwtVerify, SignJWT } from "jose";
import { z } from "zod";

// ── /compute canon ──────────────────────────────────────────────────────────
// GET /compute applies SHA-256 to COMPUTE_SEED n times and returns the
// lowercase-hex digest. n must be an integer in [1, COMPUTE_MAX_ROUNDS]; above
// the cap it is clamped (bounds per-request CPU work). The seed must equal the
// conformance runner's $sha256chain seed (benchmark/internal/conformance).
export const COMPUTE_SEED = "benchmark";
export const COMPUTE_MAX_ROUNDS = 1_000_000;

// ── /jwt/sign canon ─────────────────────────────────────────────────────────
// GET /jwt/sign issues an HS256 token with these fixed claims plus a dynamic iat
// and exp (= iat + JWT_TTL_SECONDS), signed with the shared JWT_SECRET.
export const JWT_SUBJECT = "1234567890";
export const JWT_NAME = "John Doe";
export const JWT_ADMIN = true;
export const JWT_TTL_SECONDS = 3600; // canon TTL: 1 hour

// The five canon claims echoed by /jwt/verify, in the exact shape the strict
// contract body assertion expects (no extra keys).
export type WebTokenClaims = {
  sub: string;
  name: string;
  admin: boolean;
  iat: number;
  exp: number;
};

const secretKey = (secret: string): Uint8Array => new TextEncoder().encode(secret);

// Sign the canon token. exp is pinned to iat + JWT_TTL_SECONDS so the pair is
// exact (matches the other languages' 1-hour TTL), not merely "now-ish".
export async function signToken(secret: string): Promise<string> {
  const iat = Math.floor(Date.now() / 1000);
  return new SignJWT({ name: JWT_NAME, admin: JWT_ADMIN })
    .setProtectedHeader({ alg: "HS256", typ: "JWT" })
    .setSubject(JWT_SUBJECT)
    .setIssuedAt(iat)
    .setExpirationTime(iat + JWT_TTL_SECONDS)
    .sign(secretKey(secret));
}

// Verify a token cryptographically and echo the canon claims. The algorithm is
// pinned to HS256 (rejects alg:none / algorithm-confusion) and exp is required
// and validated (jose throws JWTExpired on an expired token) — mirrors Go's
// jwt.WithValidMethods(["HS256"]) + jwt.WithExpirationRequired(). Throws on any
// invalid token (bad signature, malformed, expired, missing exp); the caller
// turns that into a 401.
export async function verifyToken(secret: string, token: string): Promise<WebTokenClaims> {
  const { payload } = await jwtVerify(token, secretKey(secret), {
    algorithms: ["HS256"],
    requiredClaims: ["exp"]
  });
  // Reconstruct the exact five-claim shape from the verified payload so the echo
  // never carries an unexpected key (strict body match) and is fully typed.
  return {
    sub: String(payload.sub),
    name: String(payload.name),
    admin: Boolean(payload.admin),
    iat: Number(payload.iat),
    exp: Number(payload.exp)
  };
}

// ── /compute ────────────────────────────────────────────────────────────────
// Parse the raw `n` query value into a clamped round count, or null when it is
// missing / non-integer / < 1 / out of range (validate-at-boundary, house
// "invalid n" 400). Semantics mirror Go's strconv.Atoi exactly: ASCII digits
// with an optional leading sign only (the `\d` class is ASCII, so underscores
// and Unicode digits are rejected and no trimming happens), parsed as a signed
// 64-bit integer. BigInt does the range check because Number loses precision
// past 2^53 — e.g. 9300000000000000000 (> i64::MAX, < u64::MAX) must be rejected,
// which a Number-based check would wrongly clamp and accept.
const COMPUTE_I64_MAX = 9223372036854775807n;
export function parseComputeRounds(raw: string | null | undefined): number | null {
  if (raw == null || !/^[+-]?\d+$/.test(raw)) return null;
  const n = BigInt(raw);
  if (n < 1n || n > COMPUTE_I64_MAX) return null;
  return n > BigInt(COMPUTE_MAX_ROUNDS) ? COMPUTE_MAX_ROUNDS : Number(n);
}

// Apply SHA-256 to the seed bytes `rounds` times and return the lowercase-hex
// digest. Web Crypto's subtle.digest is async, useless for a tight million-round
// chain; node:crypto's synchronous createHash is the right tool and is supported
// identically on Node, Bun, and Deno. Run inline on the request path (no worker
// offload) so the single-process shape matches every other language's /compute
// (fairness — typescript.md rule 21).
export function sha256Chain(rounds: number): string {
  let digest = createHash("sha256").update(COMPUTE_SEED).digest();
  for (let i = 1; i < rounds; i++) {
    digest = createHash("sha256").update(digest).digest();
  }
  return digest.toString("hex");
}

// ── /validate ───────────────────────────────────────────────────────────────
// The ~4-level POST /validate schema (contract/web.json canon). Kept module-local
// (not exported) so oxc isolated-declarations needs no annotation on it; the
// exported surface is the validate() helper, which is the shared rule set every
// server runs — the single source of truth (PLAN §3: validation rules are shared
// infrastructure, so no framework re-declares them in its own validator).
//
// Optionality mirrors the Go reference (shared/go/web ValidatePayload) exactly:
// fields Go only range-checks (age gte=0,lte=120 / total gte=0 / tags untagged)
// are zero-valued by encoding/json when omitted and still validate — zod
// `.default()` reproduces that. Fields Go marks `required` (user, id, email,
// profile, role, preferences, notifications, theme, items min=1, sku — and
// quantity, whose gte=1 rejects the zero value anyway) stay required here too.
const zValidatePayload = z.object({
  user: z.object({
    id: z.uuid(),
    email: z.email(),
    profile: z.object({
      age: z.number().int().min(0).max(120).default(0),
      role: z.enum(["admin", "user", "guest"]),
      preferences: z.object({
        theme: z.enum(["light", "dark"]),
        notifications: z.boolean()
      })
    })
  }),
  items: z
    .array(
      z.object({
        sku: z.string().min(1),
        quantity: z.number().int().min(1).max(100),
        tags: z.array(z.string()).default([])
      })
    )
    .min(1),
  total: z.number().min(0).default(0)
});

// The result of running the shared /validate rules: ok, or a per-framework
// details summary (the contract asserts $present, not an exact error count —
// zod/pydantic/validator/serde count failures differently).
export type ValidateResult = { ok: true } | { ok: false; details: string };

export function validateWebPayload(input: unknown): ValidateResult {
  const parsed = zValidatePayload.safeParse(input);
  return parsed.success ? { ok: true } : { ok: false, details: parsed.error.message };
}

File: shared/rust/src/web.rs
//! Web-suite infrastructure shared by every server that implements it (PLAN §5):
//! the `/validate` request schema (validator rules), the canon constants for
//! `/compute` and `/jwt/sign`, the HS256 sign/verify helpers (house pick:
//! `jsonwebtoken`), and the SHA-256 compute chain. Only framework-independent
//! contract values live here — the handlers stay per-framework and idiomatic
//! (PLAN §3, the idiom boundary).

use jsonwebtoken::{Algorithm, DecodingKey, EncodingKey, Header, Validation, decode, encode};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use validator::{Validate, ValidationError};

// --- Compute canon -----------------------------------------------------------

/// `GET /compute` applies SHA-256 to `COMPUTE_SEED` n times and returns the
/// lowercase-hex digest. The seed must equal the conformance runner's
/// `$sha256chain` seed (benchmark/internal/conformance).
pub const COMPUTE_SEED: &str = "benchmark";
/// n is clamped to this cap (bounds the per-request CPU work); above it the
/// response is still 200 and equals the cap-rounds chain.
pub const COMPUTE_MAX_ROUNDS: u64 = 1_000_000;

// --- JWT canon ---------------------------------------------------------------

/// `GET /jwt/sign` issues an HS256 token with these fixed claims plus a dynamic
/// `iat` and `exp` (= iat + `JWT_TTL_SECS`), signed with the shared `JWT_SECRET`.
pub const JWT_SUBJECT: &str = "1234567890";
pub const JWT_NAME: &str = "John Doe";
pub const JWT_ADMIN: bool = true;
/// Canon TTL = 1 hour (`exp = iat + 3600`).
pub const JWT_TTL_SECS: u64 = 3600;

/// The five canon claims carried by a `/jwt/sign` token and echoed by
/// `/jwt/verify`. Serializes to exactly `{sub, name, admin, iat, exp}`.
#[derive(Debug, Serialize, Deserialize)]
pub struct Claims {
    pub sub: String,
    pub name: String,
    pub admin: bool,
    pub iat: u64,
    pub exp: u64,
}

/// Sign the canon claims as an HS256 token with a fresh `iat`/`exp`.
pub fn sign_jwt(secret: &str) -> Result<String, jsonwebtoken::errors::Error> {
    let iat = jsonwebtoken::get_current_timestamp();
    let claims = Claims {
        sub: JWT_SUBJECT.to_string(),
        name: JWT_NAME.to_string(),
        admin: JWT_ADMIN,
        iat,
        exp: iat + JWT_TTL_SECS,
    };
    encode(
        &Header::new(Algorithm::HS256),
        &claims,
        &EncodingKey::from_secret(secret.as_bytes()),
    )
}

/// Verify an HS256 token against `secret`, returning the claims only when the
/// signature checks out **and** `exp` is present and unexpired. The algorithm is
/// pinned to HS256 (a token that lies about its `alg` cannot slip through) and
/// `exp` is both required and validated (a signature-valid but expired token is
/// rejected). Any failure collapses to `None` — the caller maps that to a 401.
pub fn verify_jwt(token: &str, secret: &str) -> Option<Claims> {
    let mut validation = Validation::new(Algorithm::HS256);
    validation.validate_exp = true;
    validation.set_required_spec_claims(&["exp"]);
    decode::<Claims>(
        token,
        &DecodingKey::from_secret(secret.as_bytes()),
        &validation,
    )
    .map(|data| data.claims)
    .ok()
}

// --- Compute chain -----------------------------------------------------------

/// Apply SHA-256 to `COMPUTE_SEED` `rounds` times, returning the lowercase-hex
/// digest. Callers clamp `rounds` to `COMPUTE_MAX_ROUNDS` first.
pub fn compute_chain(rounds: u64) -> String {
    let mut state = COMPUTE_SEED.as_bytes().to_vec();
    for _ in 0..rounds {
        state = Sha256::digest(&state).to_vec();
    }
    hex::encode(state)
}

// --- HTML canon --------------------------------------------------------------

/// Render the `/html` canon page: a greeting, a fruit list, and a labeled total.
/// The values are fixed cross-framework canon, so a `format!` template is the
/// honest minimal render — no engine or untrusted interpolation is involved.
pub fn render_html() -> String {
    use std::fmt::Write as _;

    let name = "Alice";
    let fruits = ["apple", "banana", "cherry"];
    let total = 42;
    let mut items = String::new();
    for fruit in fruits {
        // Writing to a String is infallible, so the Result is discarded.
        let _ = writeln!(items, "    <li>{fruit}</li>");
    }
    format!(
        "<!DOCTYPE html>\n<html lang=\"en\">\n<head><meta charset=\"utf-8\"><title>Benchmark</title></head>\n<body>\n  <h1>Hello, {name}</h1>\n  <ul>\n{items}  </ul>\n  <p>Total: {total}</p>\n</body>\n</html>\n"
    )
}

// --- /validate schema (~4 levels) --------------------------------------------

/// `POST /validate` request schema. Mirrors the cross-language canon (the Go
/// `web.ValidatePayload`, the TS zod schema, the pydantic model): a valid object
/// is `200 {"valid":true}`, any violation is `400 {"error":"validation failed"}`.
#[derive(Debug, Deserialize, Validate)]
pub struct ValidatePayload {
    #[validate(required, nested)]
    pub user: Option<ValidateUser>,
    #[validate(length(min = 1), nested)]
    pub items: Vec<ValidateItem>,
    #[validate(range(min = 0.0))]
    pub total: f64,
}

/// A UUID id, an email, and a required profile. `Serialize` is derived because
/// validator embeds the field value in its error report (`add_param`), which
/// requires every validated nested type to be serializable.
#[derive(Debug, Deserialize, Serialize, Validate)]
pub struct ValidateUser {
    #[validate(custom(function = "validate_uuid"))]
    pub id: String,
    #[validate(email)]
    pub email: String,
    #[validate(required, nested)]
    pub profile: Option<ValidateProfile>,
}

/// An age range, a role enum, and nested preferences. `age` is not required:
/// when omitted it defaults to 0 (mirroring Go's zero-value decode), which is
/// in range — an absent age validates.
#[derive(Debug, Deserialize, Serialize, Validate)]
pub struct ValidateProfile {
    #[serde(default)]
    #[validate(range(min = 0, max = 120))]
    pub age: i64,
    #[validate(custom(function = "validate_role"))]
    pub role: String,
    #[validate(required, nested)]
    pub preferences: Option<ValidatePreferences>,
}

/// The deepest level: a theme enum and a notifications flag.
#[derive(Debug, Deserialize, Serialize, Validate)]
pub struct ValidatePreferences {
    #[validate(custom(function = "validate_theme"))]
    pub theme: String,
    #[validate(required)]
    pub notifications: Option<bool>,
}

/// One line item: a sku, an in-range quantity, and a tags list (tags may be
/// empty, so it carries no presence rule).
#[derive(Debug, Deserialize, Serialize, Validate)]
pub struct ValidateItem {
    #[validate(length(min = 1))]
    pub sku: String,
    #[validate(range(min = 1, max = 100))]
    pub quantity: i64,
    pub tags: Vec<String>,
}

fn validate_uuid(value: &str) -> Result<(), ValidationError> {
    uuid::Uuid::parse_str(value)
        .map(|_| ())
        .map_err(|_| ValidationError::new("uuid"))
}

fn validate_role(value: &str) -> Result<(), ValidationError> {
    if matches!(value, "admin" | "user" | "guest") {
        Ok(())
    } else {
        Err(ValidationError::new("role"))
    }
}

fn validate_theme(value: &str) -> Result<(), ValidationError> {
    if matches!(value, "light" | "dark") {
        Ok(())
    } else {
        Err(ValidationError::new("theme"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // The dev-default shared secret (matches env.rs and the contract harness).
    const SECRET: &str = "benchmarks-shared-jwt-secret-dev-default";
    // Static tokens from contract/web.json. BAD_SIG is signed with a throwaway
    // secret; EXPIRED is signed with the dev-default secret but expired in 2020.
    const BAD_SIG: &str = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhZG1pbiI6dHJ1ZSwiZXhwIjo0MTAyNDQ0ODAwLCJpYXQiOjE3MzU2ODk2MDAsIm5hbWUiOiJKb2huIERvZSIsInN1YiI6IjEyMzQ1Njc4OTAifQ.J75FiSXpAhQxN9jiUjBHADeu_su1WJnZjJqDXI4aOWw";
    const EXPIRED: &str = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhZG1pbiI6dHJ1ZSwiZXhwIjoxNTc3ODQwNDAwLCJpYXQiOjE1Nzc4MzY4MDAsIm5hbWUiOiJKb2huIERvZSIsInN1YiI6IjEyMzQ1Njc4OTAifQ.8XxPN0yJufkzy8TdEspyV-GqR1b1MF8aW_YVERdoRic";

    const VALID_BODY: &str = r#"{
        "user": {
            "id": "3f1a2b3c-4d5e-6f70-8192-a3b4c5d6e7f8",
            "email": "alice@conformance-suite.com",
            "profile": {
                "age": 30,
                "role": "admin",
                "preferences": { "theme": "dark", "notifications": true }
            }
        },
        "items": [
            { "sku": "SKU-1", "quantity": 2, "tags": ["new", "featured"] },
            { "sku": "SKU-2", "quantity": 100, "tags": [] }
        ],
        "total": 42.5
    }"#;

    const INVALID_BODY: &str = r#"{
        "user": {
            "id": "not-a-uuid",
            "email": "not-an-email",
            "profile": {
                "age": 200,
                "role": "superuser",
                "preferences": { "theme": "neon", "notifications": true }
            }
        },
        "items": [{ "sku": "SKU-1", "quantity": 0, "tags": ["x"] }],
        "total": -5
    }"#;

    #[test]
    fn sign_then_verify_roundtrips() {
        let token = sign_jwt(SECRET).expect("sign");
        let claims = verify_jwt(&token, SECRET).expect("freshly signed token verifies");
        assert_eq!(claims.sub, JWT_SUBJECT);
        assert_eq!(claims.name, JWT_NAME);
        assert!(claims.admin);
        assert_eq!(claims.exp, claims.iat + JWT_TTL_SECS);
    }

    #[test]
    fn rejects_bad_signature() {
        assert!(verify_jwt(BAD_SIG, SECRET).is_none());
    }

    #[test]
    fn rejects_expired_token() {
        assert!(verify_jwt(EXPIRED, SECRET).is_none());
    }

    #[test]
    fn rejects_malformed_token() {
        assert!(verify_jwt("not-a-jwt", SECRET).is_none());
    }

    #[test]
    fn rejects_wrong_secret() {
        let token = sign_jwt(SECRET).expect("sign");
        assert!(verify_jwt(&token, "a-different-secret").is_none());
    }

    #[test]
    fn compute_chain_known_vectors() {
        // SHA-256 applied n times to the seed bytes "benchmark", lowercase hex.
        assert_eq!(
            compute_chain(1),
            "0e89820860c342f2c7ec694d144023b10301c2accdd078cb5167a06d0c3d5bcc"
        );
        assert_eq!(
            compute_chain(2),
            "b061accddb3b47684b3bd36291c9b219cfbd43bb074f72251f6574b073425003"
        );
    }

    #[test]
    fn validate_accepts_valid_object() {
        let payload: ValidatePayload = serde_json::from_str(VALID_BODY).expect("decode");
        assert!(payload.validate().is_ok());
    }

    #[test]
    fn validate_rejects_invalid_object() {
        let payload: ValidatePayload = serde_json::from_str(INVALID_BODY).expect("decode");
        assert!(payload.validate().is_err());
    }

    /// Decode `VALID_BODY`, apply `patch` to the parsed JSON, then decode+validate.
    /// `Err(true)` = decode failed; `Err(false)` = validation failed; `Ok(())` = valid.
    fn decode_and_validate(patch: impl FnOnce(&mut serde_json::Value)) -> Result<(), bool> {
        let mut body: serde_json::Value = serde_json::from_str(VALID_BODY).expect("base body");
        patch(&mut body);
        let payload: ValidatePayload = serde_json::from_value(body).map_err(|_| true)?;
        payload.validate().map_err(|_| false)
    }

    #[test]
    fn validate_accepts_omitted_age() {
        // age is not required (canon): omitted → defaults to 0, which is in range.
        let result = decode_and_validate(|body| {
            body["user"]["profile"]
                .as_object_mut()
                .expect("profile object")
                .remove("age");
        });
        assert_eq!(result, Ok(()));
    }

    #[test]
    fn validate_rejects_empty_items() {
        let result = decode_and_validate(|body| body["items"] = serde_json::json!([]));
        assert_eq!(result, Err(false), "empty items must fail validation");
    }

    #[test]
    fn validate_rejects_empty_sku() {
        let result = decode_and_validate(|body| body["items"][0]["sku"] = "".into());
        assert_eq!(result, Err(false), "empty sku must fail validation");
    }
}
